server/src/utils/media.ts

```diff
@@ -643,14 +643,67 @@ export class QSVConfig extends BaseHWConfig {
   }
 }
 
+export class QsvHwDecodeConfig extends QsvSwDecodeConfig {
+  getBaseInputOptions() {
+    if (this.devices.length === 0) {
+      throw new Error('No QSV device found');
+    }
+
+    const options = ['-hwaccel qsv', '-async_depth 4', '-threads 1'];
+    const hwDevice = this.getPreferredHardwareDevice();
+    if (hwDevice) {
+      options.push(`-qsv_device ${hwDevice}`);
+    }
+
+    return options;
+  }
+
+  getFilterOptions(videoStream: VideoStreamInfo) {
+    const options = [];
+    if (this.shouldScale(videoStream) || !this.shouldToneMap(videoStream)) {
+      let scaling = `scale_qsv=${this.getScaling(videoStream)}:async_depth=4:mode=hq`;
+      if (!this.shouldToneMap(videoStream)) {
+        scaling += ':format=nv12';
+      }
+      options.push(scaling);
+    }
+
+    options.push(...this.getToneMapping(videoStream));
+    return options;
+  }
+
+  getToneMapping(videoStream: VideoStreamInfo): string[] {
+    if (!this.shouldToneMap(videoStream)) {
+      return [];
+    }
+
+    const colors = this.getColors();
+    const tonemapOptions = [
+      'desat=0',
+      'format=nv12',
+      `matrix=${colors.matrix}`,
+      `primaries=${colors.primaries}`,
+      'range=pc',
+      `tonemap=${this.config.tonemap}`,
+      `transfer=${colors.transfer}`,
+    ];
+
+    return [
+      'hwmap=derive_device=opencl',
+      `tonemap_opencl=${tonemapOptions.join(':')}`,
+      'hwmap=derive_device=vaapi:reverse=1',
```

Comment:
`hwmap=derive_device=vaapi:reverse=1`
This will be not allowed in the future. It should be `hwmap=derive_device=qsv:reverse=1`
The qsv encoder's ability to accept vaapi fmt is experimental and was decided by Intel not to be promoted to upstream.

Reply:
I used `vaapi` there because `qsv` throws an error:
```
[AVHWDeviceContext @ 0x24e400304c0] 0.0: Intel(R) OpenCL Graphics / Intel(R) Arc(TM) Graphics
[AVHWDeviceContext @ 0x24e400304c0] Intel QSV to OpenCL mapping function found (clCreateFromVA_APIMediaSurfaceINTEL).
[AVHWDeviceContext @ 0x24e400304c0] Intel QSV in OpenCL acquire function found (clEnqueueAcquireVA_APIMediaSurfacesINTEL).
[AVHWDeviceContext @ 0x24e400304c0] Intel QSV in OpenCL release function found (clEnqueueReleaseVA_APIMediaSurfacesINTEL).
[AVHWFramesContext @ 0x24e401d4740] The hardware pixel format 'vaapi' is not supported by the device type 'QSV'
[Parsed_hwmap_2 @ 0x24e40173000] Failed to initialise target frames context: -38.
[Parsed_hwmap_2 @ 0x24e40173000] Failed to configure output pad on Parsed_hwmap_2
Error reinitializing filters!
Failed to inject frame into filter network: Function not implemented
Error while processing the decoded data for stream #0:0
[AVIOContext @ 0x24e401907c0] Statistics: 0 bytes written, 0 seeks, 0 writeouts
Terminating demuxer thread 0
[AVIOContext @ 0x24e401902c0] Statistics: 1982895 bytes read, 2 seeks
Conversion failed!
```
Is there something else I would need to change to make this work? The full command is this:
```
ffmpeg -hwaccel qsv -async_depth 4 -threads 1 -i HDR.mp4 \
    -c:v h264_qsv -c:a copy -movflags faststart -fps_mode passthrough -map 0:0 -bf 7 -refs 5 -g 256 -v verbose \
    -vf hwmap=derive_device=opencl,tonemap_opencl=desat=0:format=nv12:matrix=bt709:primaries=bt709:range=pc:tonemap=hable:transfer=bt709,hwmap=derive_device=qsv:reverse=1 \
    -preset 7 -global_quality 23 SDR.mp4
```

Reply:
I think you have to specify the `format=qsv` to enable the indirect hwmap.

`hwmap=derive_device=qsv:reverse=1,format=qsv`

Reply:
Thanks, that worked! I'll change it to use `qsv`.